Migrate OCIRepository controller to runtime/secrets

cappyzawa opened the pull request:
## Summary

This PR migrates OCIRepository authentication to use runtime/secrets API as part of fluxcd/flux2#5433.

The migration moves TLS configuration from internal/tls to runtime/secrets.TLSConfigFromSecretRef and ServiceAccount processing to secrets.PullSecretsFromServiceAccountRef, providing consistent authentication handling across all source-controller components.

matheuscscp commented:
Proxy is missing :grin:

matheuscscp commented:
```diff
diff --git a/internal/controller/ocirepository_controller.go b/internal/controller/ocirepository_controller.go
index 97ae1b1..af54e33 100644
--- a/internal/controller/ocirepository_controller.go
+++ b/internal/controller/ocirepository_controller.go
@@ -354,14 +354,21 @@ func (r *OCIRepositoryReconciler) reconcileSource(ctx context.Context, sp *patch
 		return sreconcile.ResultEmpty, e
 	}
 
-	proxyURL, err := r.getProxyURL(ctx, obj)
-	if err != nil {
-		e := serror.NewGeneric(
-			fmt.Errorf("failed to get proxy address: %w", err),
-			sourcev1.AuthenticationFailedReason,
-		)
-		conditions.MarkTrue(obj, sourcev1.FetchFailedCondition, e.Reason, "%s", e)
-		return sreconcile.ResultEmpty, e
+	var proxyURL *url.URL
+	if obj.Spec.ProxySecretRef != nil {
+		var err error
+		proxyURL, err = secrets.ProxyURLFromSecretRef(ctx, r.Client, types.NamespacedName{
+			Name:      obj.Spec.ProxySecretRef.Name,
+			Namespace: obj.GetNamespace(),
+		})
+		if err != nil {
+			e := serror.NewGeneric(
+				fmt.Errorf("failed to get proxy address: %w", err),
+				sourcev1.AuthenticationFailedReason,
+			)
+			conditions.MarkTrue(obj, sourcev1.FetchFailedCondition, e.Reason, "%s", e)
+			return sreconcile.ResultEmpty, e
+		}
 	}
 
 	if _, ok := keychain.(soci.Anonymous); obj.Spec.Provider != "" && obj.Spec.Provider != sourcev1.GenericOCIProvider && ok {
```

Then `func (r *OCIRepositoryReconciler) getProxyURL()` can be removed

stefanprodan commented:
What about https://github.com/fluxcd/source-controller/blob/173a1cccc5cda6c1a2ac1bd74b092b4d682f0ab0/internal/tls/config.go#L124-L126

And https://github.com/fluxcd/source-controller/blob/173a1cccc5cda6c1a2ac1bd74b092b4d682f0ab0/internal/tls/config.go#L147-L154

matheuscscp commented:
@stefanprodan Would it make sense to add those to `runtime/secrets`?

stefanprodan commented:
We should get rid of `internal/tls/config.go` and looks like 1.2 is now the min default https://go-review.googlesource.com/c/go/+/541516

stefanprodan commented:
> Would it make sense to add those to runtime/secrets and have them everywhere?

Yes for the min TLS version

> Not sure how to go about the url, since the inputs for the library TLS config are just the Secret object or reference

I think the URL validation happen for OCIRepo and HelmRepo OCI no matter if TLS is set or not. For sure it doesn't belong in `runtime/secrets`

matheuscscp commented:
`ImageRepository` did not have this `tlsConf.ServerName = u.Hostname()` before the migration to `runtime/secrets`. What's the effect of this assignment? If we remove this are we breaking any setups?

```go

	// ServerName is used to verify the hostname on the returned
	// certificates unless InsecureSkipVerify is given. It is also included
	// in the client's handshake to support virtual hosting unless it is
	// an IP address.
	ServerName string
```

This tells me we would breaking a feature:

```
It is also included in the client's handshake to support virtual hosting unless it is an IP address.
```

cappyzawa commented:
@stefanprodan Thanks for pointing out those missing parts!

Regarding **TLS MinVersion**: Go 1.20+ defaults to TLS 1.2 as you mentioned, so I believe the explicit setting is no longer necessary. I didn't add it to maintain consistency with the Go defaults.

Regarding **ServerName**: You're right - this was a critical regression in the migration from `internal/tls` to `runtime/secrets`. Without ServerName, TLS handshakes fail with certificate mismatch errors in virtual hosting environments. I've fixed this in 26abfa9a for OCIRepository.

Additionally, I discovered that #1849 had the same ServerName issue, which I've also addressed in 06b3d72.

cappyzawa commented:
Regarding removing `internal/tls/config.go`: I agree! Once we migrate the Bucket controller to `runtime/secrets` in a follow-up PR, there should be no more dependencies on `internal/tls/config.go`, and we can remove it completely at that time.

matheuscscp commented:
For consistency we should probably also implement `ServerName` for `ImageRepository` since both OCI APIs in source-controller have this feature

matheuscscp reviewed (approved):
LGTM! 🚀

matheuscscp reviewed (approved):
LGTM! 🚀

cappyzawa commented:
Got the approvals, so I'll go ahead and squash!

stefanprodan reviewed (approved):
LGTM
